update/v1alpha1/types_clusterversion_progress_insight.go

```diff
@@ -0,0 +1,263 @@
+	// +resource name (because the rest is implied by status insight type). However, because we use resource references in
+	// +many places and this API is intended to be consumed by clients, not produced, consistency seems to be more valuable
+	// +than type safety for producers.
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.group == 'config.openshift.io' && self.resource == 'clusterversions'",message="resource must be a clusterversions.config.openshift.io resource"
+	Resource ResourceRef `json:"resource"`
+
+	// assessment is the assessment of the control plane update process. Valid values are: Unknown, Progressing, Completed, Degraded
+	// +required
+	Assessment ControlPlaneAssessment `json:"assessment"`
+
+	// versions contains the original and target versions involved in the update
+	// +required
+	Versions ControlPlaneUpdateVersions `json:"versions"`
+
+	// completion is a percentage of the update completion (0-100)
+	// +required
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=100
+	Completion int32 `json:"completion"`
+
+	// startedAt is the time when the update started
+	// +required
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	StartedAt metav1.Time `json:"startedAt"`
+
+	// completedAt is the time when the update completed
+	// +optional
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	CompletedAt *metav1.Time `json:"completedAt,omitempty"`
+
+	// estimatedCompletedAt is the estimated time when the update will complete
+	// +optional
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	EstimatedCompletedAt *metav1.Time `json:"estimatedCompletedAt,omitempty"`
+}
+
+// ControlPlaneAssessment is the assessment of the control plane update process
+// +kubebuilder:validation:Enum=Unknown;Progressing;Completed;Degraded
+type ControlPlaneAssessment string
+
+const (
+	// Unknown means the update status and health cannot be determined
+	ControlPlaneAssessmentUnknown ControlPlaneAssessment = "Unknown"
+	// Progressing means the control plane is updating and no problems or slowness are detected
+	ControlPlaneAssessmentProgressing ControlPlaneAssessment = "Progressing"
+	// Completed means the control plane successfully completed updating and no problems are detected
+	ControlPlaneAssessmentCompleted ControlPlaneAssessment = "Completed"
+	// Degraded means the process of updating the control plane suffers from an observed problem
+	ControlPlaneAssessmentDegraded ControlPlaneAssessment = "Degraded"
+)
+
+// ControlPlaneUpdateVersions contains the original and target versions of the upgrade
+type ControlPlaneUpdateVersions struct {
+	// previous is the version of the control plane before the update. When the cluster is being installed
+	// for the first time, the version will have a placeholder value '<none>' and carry 'Installation' metadata
```

Comment:
nit: if the cluster history is:

1. Completed vA.
2. Partial vB.
3. Partial vC.

it's not clear to me from these docs whether the value would be vA or vB.  [The client-side code suggests vB][1], in which case maybe wording like:

> previous is the version the control plane was attempting to reconcile before the update.

or some such that makes it clear that it's "the previous target, regardless of completion" and not "the most recent completed target".

[1]: https://github.com/openshift/oc/blob/fe18ae25a123e896fd218210073141f7f5811465/pkg/cli/admin/upgrade/status/controlplane.go#L416-L418

Reply:
Improved the godoc, hopefully